common/futures-semaphore/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+//! to a shared resource. Before accessing the shared resource, callers must
+//! acquire a permit from the semaphore. Once the permit is acquired, the caller
+//! then enters the critical section. When the caller is finished, they drop the
+//! permit to release it back to the semaphore.
+//!
+//! `Semaphore` is futures-aware. Acquiring a [`Permit`] is an async function,
+//! which will yield to the futures executor if no permits are available and be
+//! woken up when one becomes available.
+//!
+//! In contrast, acquiring a permit from a [`std::sync::Mutex`] +
+//! [`std::sync::Condvar`] style semaphore will block the entire OS thread, which
+//! could potentially deadlock the futures runtime if enough tasks are waiting on
+//! the semaphore.
+
+#![feature(async_await)]
+
+use futures01::{future::Future as Future01, Async as Async01, Poll as Poll01};
+use futures03::compat::Future01CompatExt;
+use std::sync::Arc;
+use tokio_sync::semaphore::{AcquireError, Permit as TokioPermit, Semaphore as TokioSemaphore};
+
+/// The wrapped tokio_sync [`Semaphore`](TokioSemaphore) and total permit capacity.
+#[derive(Debug)]
+struct Inner {
+    semaphore: TokioSemaphore,
+    capacity: usize,
+}
+
+/// A futures-aware semaphore.
+#[derive(Clone, Debug)]
+pub struct Semaphore {
+    inner: Arc<Inner>,
+}
+
+/// A permit acquired from a semaphore, allowing access to a shared resource.
+/// Dropping a `Permit` will release it back to the semaphore.
+#[derive(Debug)]
+pub struct Permit {
+    inner: Arc<Inner>,
+    permit: TokioPermit,
+}
+
+impl Semaphore {
+    /// Create a new semaphore with `capacity` number of available permits.
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            inner: Arc::new(Inner {
+                semaphore: TokioSemaphore::new(capacity),
+                capacity,
+            }),
+        }
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.inner.capacity
+    }
+
+    pub fn available_permits(&self) -> usize {
+        self.inner.semaphore.available_permits()
+    }
+
+    pub fn is_idle(&self) -> bool {
+        self.available_permits() == self.capacity()
+    }
+
+    pub fn is_full(&self) -> bool {
+        self.available_permits() == 0
+    }
+
+    /// Acquire an available permit from the semaphore, blocking until none are
+    /// available.
+    pub async fn acquire(&self) -> Permit {
+        let permit = Permit {
+            inner: Arc::clone(&self.inner),
+            permit: TokioPermit::new(),
+        };
+
+        PermitFuture01::new(permit)
+            .compat()
+            .await
+            // The TokioSemaphore is not dropped yet and our wrapper never calls
+            // .close(), so the TokioSemaphore can never be closed unless all
+            // references, including this &self, have been dropped.
+            .expect("TokioSemaphore is never closed")
+
+        // TODO(philiphayes): use this when tokio updates to v0.2
```

Comment:
Personally, I'd remove this kind of commented out code which is saved for later, just to keep the code cleaner and easier to read.